File: Application/src/main/java/com/shareyourproxy/api/rx/JustObserver.java
package com.shareyourproxy.api.rx;

import android.util.Log;

import rx.Observer;
import timber.log.Timber;

/**
 * This abstraction simply logs all errors to the command prompt.
 */
public abstract class JustObserver<T> implements Observer<T> {

    @Override
    public void onCompleted() {
        Timber.v("%1$s completed", this.toString());
        complete();
    }

    @Override
    public void onError(Throwable e) {
        Timber.e(Log.getStackTraceString(e));
        error(e);
    }

    @Override
    public void onNext(T t) {
        Timber.v("%1$s onNext obj: %2$s", this.toString(), t.toString());
        next(t);
    }

    public abstract void next(T t);

    public void error(Throwable e) {
    }

    public void complete() {
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/JustObserver.kt
package com.shareyourproxy.api.rx

import android.util.Log
import rx.Observer
import timber.log.Timber

/**
 * This abstraction simply logs all errors to the command prompt.
 */
internal abstract class JustObserver<T>(private val clazz: Class<*>) : Observer<T> {

    override fun onCompleted() {
        Timber.v("${clazz.simpleName} completed")
        complete()
    }

    override fun onError(e: Throwable) {
        Timber.e("${clazz.simpleName}: ${Log.getStackTraceString(e)}")
        error(e)
    }

    override fun onNext(t: T) {
        Timber.v("${clazz.simpleName} onNext obj: ${t.toString()}")
        next(t)
    }

    abstract fun next(t: T)

    open fun error(e: Throwable) {
    }

    open fun complete() {
    }
}
